docs/design/secure-config.md

```diff
@@ -0,0 +1,219 @@
+
+```proto
+message Metadata {
+  // This is used to distinguish between copies and moves.
+  string path = 1;
+}
+```
+
+The function to load repository configuration, will roughly speaking, look like:
+```rust
+enum ConfigLoadError {
+    NoRepoId,
+    NoConfig,
+    PathMismatch,
+}
+
+fn load_repo_config_path(repo: Path) -> Result<PathBuf, ConfigLoadError> {
+    let Ok(repo_id) = std::fs::read_to_string(repo.join("repo-id")) else return Err(NoRepoId);
+    let repo_config_dir = config_dir.join("repos").join(repo_id);
+    let Ok(metadaata) = Metadata::decode(std::fs::read(repo_config_dir.join("metadata.binpb"))) else return Err(NoConfig);
+    if metadata.path != repo {
+        return Err(PathMismatch)
+    }
+    Ok(repo_config_dir.join("config.toml"))
+}
+```
+
+### No repo-ID
+All new repos should have repo IDs, so this should mean that we have a repo created by an old version of jj.
+
+To preserve backwards compatibility, we will introduce a period of auto-migration. The current plan is 12 jj versions (approximately 1 year). During this period, if a repo ID has not yet been generated, we will silently perform the following (order matters, to ensure failure halfway through doesn’t affect things):
+
+1. Generate a repo ID `abc123`
+2. Create `$HOME/.config/jj/repo/abc123/metadata.binpb`
+3. Create `$HOME/.config/jj/repo/abc123/config.toml` as a copy of the original config file
+4. Atomically generate a Repo-ID file containing `abc123`
+5. Remove the original config file
+6. Symlink the original config file to the new config file
+
+After the migration period is over, we will:
+* Add comments at the start of every line in step 3
+* Print a warning to the user that their config has been migrated, but commented out, and that they need to run `jj config edit` to get the config back.
+
+### No Config
+This could occur, for example, if the user created a repository in linux, rebooted into windows on the same computer, and attempted to access that repo.
+
+In this event, the user probably expects their config to be attached to the repository, and they expect it to still work on linux, so we will:
+* Create a new repo with the same repo ID (to ensure that the config still works on windows)
+* Print a warning that if there was any per-repo configuration, it is no longer available.
+
+### Path mismatch
+#### Secure
+If we have a path mismatch, one of the following things have occurred:
+
+* The repo has been moved
+* The repo has been copied
+* The user is a victim of a replay attack where the user replays the repo id
+
+We can distinguish between the first two by attempting to check whether the old path still exists on disk.
+
+* If the old path does not exist on disk, the repo has been moved, so we need to update the path in `metadata.binpb`
+* If the old path does exist on disk, the repo has been copied, so we need to generate a new repo ID, and create a copy of the directory `.config/jj/OLD_REPO_ID` for the new repo ID
```

Comment:
The old path might be temporary unavailable, but in that case, the repository at the old location will be considered "copied". So I think this is good.